suzieq/engines/pandas/ospf.py

```diff
@@ -280,41 +290,42 @@ def aver(self, **kwargs):
                     if (x["hostname_y"] is not np.nan and
                         len(x['hostname_y']) != 1) else [], axis=1))
 
-        # Now  peering match
-        lldp_df = self._get_table_sqobj('lldp').get(
-            namespace=kwargs.get('namespace', []), columns=['default'])
+        passed_df = ospf_df \
+            .query('adjState.isin(["full", "passive"])') \
+            .reset_index(drop=True)
+        failed_df = ospf_df \
+            .query('~adjState.isin(["full", "passive"]) and '
+                   'ifState != "down"') \
+            .reset_index(drop=True)
+        ifdown_df = ospf_df \
+            .query('~adjState.isin(["full", "passive"]) and '
+                   'ifState == "down"') \
+            .reset_index(drop=True)
+        ifdown_df['assertReason'] += \
+            ifdown_df.ifState.apply(lambda x: ['ifdown'])
+
+        # Validate loopback interfaces are of the right network type
+        passed_df['assertReason'] += passed_df.apply(
+            lambda x: ['network type !loopback']
+            if (any(ele in x.ifname for ele in ['lo', 'oopbac']) and
+                x.networkType != "loopback") else [], axis=1)
+
+        if failed_df.empty:
+            return self._create_aver_result([passed_df, ifdown_df], result)
+
+        lldp_df = self._get_lldp_peer(failed_df)
         if lldp_df.empty:
-            ospf_df = ospf_df[~(ospf_df.ifname.str.contains('loopback') |
-                                ospf_df.ifname.str.contains('Vlan'))]
-            ospf_df['assertReason'] = 'No LLDP peering info'
-            ospf_df['result'] = 'fail'
-            return ospf_df[['namespace', 'hostname', 'vrf', 'ifname',
-                            'assertReason', 'result']]
-
-        # Create a single massive DF with fields populated appropriately
-        use_cols = [
-            "namespace",
-            "routerId",
-            "hostname",
-            "vrf",
-            "ifname",
-            "helloTime",
-            "deadTime",
-            "passive",
-            "ipAddress",
-            "areaStub",
-            "isUnnumbered",
-            "networkType",
-            "area",
-            "timestamp",
-            "lldpIfname"
-        ]
+            failed_df['assertReason'] = 'No LLDP peering info'
+            return self._create_aver_result([passed_df, ifdown_df, failed_df],
+                                            result)
 
         # In case of Junos, the OSPF interface name is a subif of
         # the interface name. So, create a new column with the
         # orignal interface name.
-        ospf_df['lldpIfname'] = ospf_df['ifname'].str.split('.').str[0]
-        int_df = ospf_df[use_cols].merge(
+        failed_df['lldpIfname'] = failed_df['ifname'].str.split('.').str[0]
+        use_cols = failed_df.columns
+
+        int_df = failed_df[use_cols].merge(
```

Comment:
If use_cols == failed_df.columns, I don't think you need to do `failed_df[use_cols]` right?

Reply:
True enough